### ISSUE TITLE
musl-cross-make toolchain can not make ATLAS

### ISSUE THREAD

**Issue by ximik69** [richfelker/musl-cross-make#33]
I have compiled new toolchain with fortran, as it is needed for compiling static ATLAS,gsl,lapack,OpenMPI,  and other tools for electronic structure calculations (JDFTx, Quantum Espresso,...). The thing I need is static toolchain, able to make fully static code with  SSE expensions for HPC. I'd like to mitigate problems with installation on different machines with alike CPUs or updating linux distro on same machine without need to rebuild libs ATLAS,gsl,lapack,OpenMPI and JDFTx.

Small (fully static)test software:  Xz and ZPAQ (7.15 and 7.05).
Generated toolchain makes good Xz, but fails to make working ZPAQ (7.15 and 7.05).
ZPAQ runs with --help, but crashes with segfault on trying to compress something. 

While compiling ATLAS, it complains that compiler doesn't support SSE:
"Default compiler & flags do not support SSEGEN!!".

I change only CC,CXX,F77 system variables.

My arch is x86-64, Core2DUO T5800. GCC 4.8.2, Mageia Linux  4.
Experiments with different gcc versions, distros give same result.

The most important parameters of config are given,others are default:
TARGET = x86_64-linux-musl
OUTPUT = /opt/cross
GCC_CONFIG += --enable-languages=c,c++,fortran

after making completely static toolchain with previously maked as above:
TARGET = x86_64-linux-musl
OUTPUT = /opt/cross2
 COMMON_CONFIG += CC="/opt/cross/bin/x86_64-linux-musl-gcc -static --static" CXX="/opt/cross/bin/x86_64-linux-musl-g++ -static --static"
GCC_CONFIG += --enable-languages=c,c++,fortran,lto

results don't change.

The configs, script for building ATLAS and ATLAS build output are in attachments.

Could you say me what am I doing wrong? And what should I do for compiling correct toolchain+ATLAS?

Thank you in advance, Igor.

**Comment by richfelker** [comment]
I don't see the attachment. There's probably something wrong in ATLAS's configure script in detecting sse; seeing the script and log of where it failed is going to be necessary to determine why.

**Comment by ximik69** [comment]
[config.mak.txt](https://github.com/richfelker/musl-cross-make/files/1128108/config.mak.txt) = Initial config for musl-cross-make.
[config (2).mak.txt](https://github.com/richfelker/musl-cross-make/files/1128112/config.2.mak.txt)= config for musl-cross-make using previously made dynamic toolchain for production of static toolchain.

[atlas.sh.txt](https://github.com/richfelker/musl-cross-make/files/1128117/atlas.sh.txt)=script for building ATLAS

[ATLAS_build.zip](https://github.com/richfelker/musl-cross-make/files/1128138/ATLAS_build.zip)=output of ATLAS building script.

**Comment by ximik69** [comment]
Dear richfelker, thanks a lot for such a quick reply. I had to change browser for attaching files.

**Comment by richfelker** [comment]
It's not clear exactly what the log is and how it relates to the configure failure you mentioned, but I suspect your problem is that the software assuming SSE3, which is not part of the baseline x86_64 ISA. Adding -msse3 or -march=[something appropriate] to CFLAGS or CC might help. I don't think this is an mcm issue.

**Comment by rofl0r** [comment]
as far as zpaq segfault is concerned, i opened an issue upstream but it was discarded since i didnt contact the author directly: https://github.com/zpaq/zpaq/pull/14

**Comment by ximik69** [comment]
Dear Rich Felker,
thanks a lot for your hint with -msse3 key.
Script is here:
[atlas.sh.txt](https://github.com/richfelker/musl-cross-make/files/1132076/atlas.sh.txt)

 It made huge step ahead. Now ATLAS compiles successfully and passes 'make check'. However solving one question raised two new ones.
1) ATLAS failed to pass 'make ptcheck' due to linkier errors:
see 
[_ptcheck.txt](https://github.com/richfelker/musl-cross-make/files/1132078/_ptcheck.txt)


/opt/cross2/bin/x86_64-linux-musl-gfortran -static --static -msse3 -static --static -msse3 -m64 -o xsqrtst_pt sqrtst_pt.o \
                   /src/ATLAS_build/lib/libtstatlas.a /src/ATLAS_build/lib/libptlapack.a /src/ATLAS_build/lib/libptcblas.a /src/ATLAS_build/lib/libptf77blas.a \
                   /src/ATLAS_build/lib/libatlas.a -lpthread -lm
sqrtst_pt.o: In function `lqtest':
qrtst.c:(.text+0xd14): undefined reference to `ATL_C2Fsormlq'
sqrtst_pt.o: In function `qltest':
qrtst.c:(.text+0x1401): undefined reference to `ATL_C2Fsormql'
sqrtst_pt.o: In function `rqtest':
qrtst.c:(.text+0x1ac9): undefined reference to `ATL_C2Fsormrq'
sqrtst_pt.o: In function `qrtest':
qrtst.c:(.text+0x2162): undefined reference to `ATL_C2Fsormqr'
/src/ATLAS_build/lib/libptlapack.a(ATL_stgelqf.o): In function `ATL_stgelqf':
ATL_gelqf.c:(.text+0x17f): undefined reference to `clapack_ilaenv'
/src/ATLAS_build/lib/libptlapack.a(ATL_stgeqlf.o): In function `ATL_stgeqlf':
ATL_geqlf.c:(.text+0x15c): undefined reference to `clapack_ilaenv'
/src/ATLAS_build/lib/libptlapack.a(ATL_stgeqrf.o): In function `ATL_stgeqrf':
ATL_geqrf.c:(.text+0x15c): undefined reference to `clapack_ilaenv'
/src/ATLAS_build/lib/libptlapack.a(ATL_stgerqf.o): In function `ATL_stgerqf':
ATL_gerqf.c:(.text+0x17f): undefined reference to `clapack_ilaenv'
collect2: error: ld returned 1 exit status


Googling and reading ERRATA gave no obvious ideas. 
Should I install LAPACK first or do something else?

2)after 'make time' timings are quite sad.
See 
[_time.txt](https://github.com/richfelker/musl-cross-make/files/1132088/_time.txt)

                    single precision                  double precision
            ********************************   *******************************
                  real           complex           real           complex
            ---------------  ---------------  ---------------  ---------------
Benchmark   Refrenc Present  Refrenc Present  Refrenc Present  Refrenc Present
=========   ======= =======  ======= =======  ======= =======  ======= =======
  kSelMM      646.0    82.9    611.2    81.3    369.1   364.6    355.4   340.7
  kGenMM      185.1    82.9    183.0    81.3    173.7    29.6    176.1    30.2
  kMM_NT      172.7    54.9    166.6    53.4    164.0    46.3    174.6    48.0
  kMM_TN      185.8    79.0    178.6    73.9    174.1    55.0    167.9    68.0
  BIG_MM      603.3    58.2    600.5    49.9    351.2   162.6    341.8   147.1
   kMV_N       74.5    90.4    147.7   147.8     43.8    53.2     88.1    43.9
   kMV_T       70.6    80.9    156.2   158.5     48.7    61.2     90.5   105.3
    kGER       39.5    37.7     81.7    50.8     20.4    25.2     40.4    44.9
make[1]: Leaving directory '/src/ATLAS_build'

But there could be little load on machine. 
I shall try '-march=native'. How do you think, could it help? or should I try another flags or even gcc version?
I had troubles with gcc4.6 with similar symptoms. It generated bad slow code. I fixed it only after changing gcc to  another version.

Much thanks, Igor.

**Comment by ximik69** [comment]
Small update:
adding -msse3 -mssse3 -march=native did not help at all. Everything remaied the same.

**Comment by ximik69** [comment]
Dear rofl0r, 
thanks for the comment. Now I understand that it is rather zpaq problem and not musl-libc or toolchain.

**Comment by rofl0r** [comment]
well, it is a problem with wrong assumptions in zpaq... if you feel motivated to get it fixed, maybe you can contact its author and forward him the text i've written in that issue together with your own findings. thanks!

**Comment by richfelker** [comment]
I don't think this issue requires any action on the musl-cross-make side so I'm closing it; please reopen if I'm mistaken.
